src/terminator/mod.rs

```diff
@@ -229,6 +229,34 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
                         (def_id, substs, Vec::new())
                     };
 
+                // FIXME(eddyb) Detect ADT constructors more efficiently.
+                if let Some(adt_def) = fn_ty.sig.skip_binder().output().ty_adt_def() {
+                    if let Some(v) = adt_def.variants.iter().find(|v| resolved_def_id == v.did) {
+                        // technically they can diverge, but only if one of their arguments diverges, so it doesn't matter
```

Comment:
Oh, that makes sense. i'm satisfied for now... Also, I just remembered this thread is attached to code specific to tuple struct constructors, which actually *can't* diverge, right? Because their return values will always be the struct that they are a constructor for.

So I guess just remove that comment? At least we fixed some other ICEs along the way. :smile: